crates/superposition/src/auth/oidc.rs

```diff
@@ -0,0 +1,327 @@
+                .and_then(|t| {
+                    t.claims(&auth.client.id_token_verifier(), &pc.nonce)
+                        .map_err(|e| log::error!("Couldn't verify claims: {e}"))
+                })?;
+            Ok(tr)
+        });
+
+    match response {
+        Ok(r) => {
+            let token = serde_json::to_string(&r).map_err(|err| {
+                log::error!("Unable to stringify data: {err}");
+                ErrorInternalServerError(format!("Unable to stringify data"))
+            })?;
+            let cookie = Cookie::build("user", token)
+                .path("/")
+                .http_only(true)
+                .max_age(Duration::days(1))
+                .finish();
+            Ok(HttpResponse::Found()
+                .cookie(cookie)
+                .insert_header(("Location", "/admin/organisations"))
+                .finish())
+        }
+        Err(()) => Ok(auth.new_redirect()),
+    }
+}
+
+#[get("/organisations")]
+async fn get_organisations(
+    auth: Data<OIDCAuthenticator>,
+    req: HttpRequest,
+) -> HttpResponse {
+    let organisations = req
+        .cookie("user")
+        .and_then(|user_cookie| {
+            auth.decode_token(user_cookie.value())
+                .map_err(|e| log::error!("Error in decoding user : {e}"))
+                .ok()
+        })
+        .map(|claims| claims.additional_claims().organisations.clone());
+
+    match organisations {
+        Some(organisations) => HttpResponse::Ok().json(serde_json::json!(organisations)),
+        None => auth.new_redirect(),
+    }
+}
+
+#[get("/organisations/switch/{organisation_id}")]
+async fn switch_organisation(
+    auth: Data<OIDCAuthenticator>,
+    req: HttpRequest,
+    path: Path<SwitchOrgParams>,
+) -> actix_web::Result<HttpResponse> {
+    let issuer_endpoint = get_from_env_unsafe::<String>("OIDC_ISSUER_ENDPOINT_FORMAT")
+        .unwrap()
+        .replace("<organisation>", &path.organisation_id);
+    let token_endpoint = get_from_env_unsafe::<String>("OIDC_TOKEN_ENDPOINT_FORMAT")
+        .unwrap()
+        .replace("<organisation>", &path.organisation_id);
+
+    let provider = auth
+        .provider_metadata
+        .clone()
+        .set_issuer(IssuerUrl::new(issuer_endpoint).map_err(|e| {
+            log::error!("Unable to create issuer url {e}");
+            ErrorBadRequest(String::from("Unable to create issuer url"))
+        })?)
+        .set_token_endpoint(TokenUrl::new(token_endpoint).ok());
+
+    let client = CoreClient::from_provider_metadata(
+        provider,
+        ClientId::new(auth.client_id.clone()),
+        Some(ClientSecret::new(auth.client_secret.clone())),
+    )
+    .set_redirect_uri(
+        RedirectUrl::new(format!("{}/", auth.base_url.clone())).map_err(|e| {
+            log::error!("Unable to create redirect url {e}");
+            ErrorInternalServerError("Unable to create redirect url")
+        })?,
+    );
+
+    let (username, switch_pass) = req
```

Comment:
Since we're already setting the `User` in the middleware, should we just add these fields as well?